File: mpi-send-recv.c
#include <stdio.h>
#include <mpi.h>

int main(int argc, char *argv[]) {

int size, rank;
double a[2],b[2];

MPI_Init(&argc,&argv);
MPI_Comm_size(MPI_COMM_WORLD,&size);
MPI_Comm_rank(MPI_COMM_WORLD,&rank);

if (rank ==0) {
   a[0]=2.0;
   a[1]=4.0;
   MPI_Send(a,2,MPI_DOUBLE,1,10,MPI_COMM_WORLD);
   MPI_Recv(b,2,MPI_DOUBLE,1,10,MPI_COMM_WORLD,MPI_STATUS_IGNORE);
}
else if (rank==1) {
   a[0]=3.0;
   a[1]=5.0;
   MPI_Recv(b,2,MPI_DOUBLE,0,10,MPI_COMM_WORLD,MPI_STATUS_IGNORE);
   MPI_Send(a,2,MPI_DOUBLE,0,10,MPI_COMM_WORLD);
}
printf("%d b[0]=%lf, b[1]=%lf \n",rank,b[0],b[1]);


MPI_Finalize();


}

File: mpi-send-recv.f90
PROGRAM avoid_lock
  USE mpi
  Implicit none
  INTEGER ierr, myid, nproc
  INTEGER status(MPI_STATUS_SIZE)
  REAL A(2), B(2)

  CALL MPI_INIT(ierr)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, myid, ierr)

  IF( myid .EQ. 0 ) THEN
     a(1) = 2.0
     a(2) = 4.0
     CALL MPI_SEND(a, 2, MPI_REAL, 1, 10, MPI_COMM_WORLD, ierr)
     CALL MPI_RECV(b, 2, MPI_REAL, 1, 11, MPI_COMM_WORLD, status, ierr)
  ELSE IF( myid .EQ. 1 ) THEN
     a(1) = 3.0
     a(2) = 5.0
     CALL MPI_SEND(a, 2, MPI_REAL, 0, 11, MPI_COMM_WORLD, ierr)
     CALL MPI_RECV(b, 2, MPI_REAL, 0, 10, MPI_COMM_WORLD, status, ierr)
  END IF
  WRITE(6,*) myid, ': b(1)=', b(1), ' b(2)=', b(2)
  CALL MPI_FINALIZE(ierr)
END program avoid_lock

